### PR TITLE
MobileNet example

### PR DESCRIPTION
**Pull request by kevin85421**
I deployed this example with RayService on KubeRay.

Screen Shot 2023-06-20 at 11 10 19 AM

### CODE REVIEW

File: mobilenet/mobilenet.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+from ray.serve.drivers import DAGDriver
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]: Could you reorder these imports to highlight the Serve parts:

```python
import numpy as np

from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.mobilenet_v2 import (
    preprocess_input,
    decode_predictions,
    MobileNetV2,
)

from ray import serve
from ray.serve.drivers import DAGDriver
```

File: mobilenet/mobilenet.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+from tensorflow.keras.preprocessing import image
+from tensorflow.keras.applications.mobilenet_v2 import preprocess_input, decode_predictions
+import numpy as np
+
+@serve.deployment()
+class ImageClassifier:
+    def __init__(self):
+        self.model = MobileNetV2(weights='imagenet')
+    # Users can send HTTP requests with an image. The classifier will return
+    # the top 1 prediction.
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]:
```suggestion
    # the top prediction.
```

File: mobilenet/mobilenet.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+from ray import serve
+from tensorflow.keras.applications.mobilenet_v2 import MobileNetV2
+from tensorflow.keras.preprocessing import image
+from tensorflow.keras.applications.mobilenet_v2 import preprocess_input, decode_predictions
+import numpy as np
+
+@serve.deployment()
+class ImageClassifier:
+    def __init__(self):
+        self.model = MobileNetV2(weights='imagenet')
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]:
```suggestion
        self.model = MobileNetV2(weights="imagenet")
```

File: mobilenet/mobilenet.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    # Users can send HTTP requests with an image. The classifier will return
+    # the top 1 prediction.
+    # Sample output: {"prediction":["n02099601","golden_retriever",0.17944198846817017]}
+    async def __call__(self, http_request):
+        request = await http_request.form()
+        image_file = await request["image"].read()
+        import tempfile
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]: Can you move this import to the top of the file?

File: mobilenet/mobilenet.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            img = image.load_img(temp_file_path, target_size=(224, 224))
+        x = image.img_to_array(img)
+        x = np.expand_dims(x, axis=0)
+        x = preprocess_input(x)
+        preds = self.model.predict(x)
+        decoded_preds = decode_predictions(preds, top=1)[0]
+        return {"prediction": decoded_preds[0]}
+
+image_classifier_handle = ImageClassifier.bind()
+graph = DAGDriver.bind(image_classifier_handle)
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]:
```suggestion
app = DAGDriver.bind(image_classifier_handle)
```

Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]: Do you need to use a `DAGDriver`? Could you directly use `app = ImageClassifier.bind()`?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work so far! I left a few comments.

File: mobilenet/mobilenet.py (hunk 5 above)
Comment (kevin85421) [reply mobilenet/mobilenet.py]: I have updated it. Let me test again.

Comment (kevin85421) [reply mobilenet/mobilenet.py]: It works.

File: mobilenet/mobilenet.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+from ray import serve
+
+@serve.deployment()
+class ImageClassifier:
+    def __init__(self):
+        self.model = MobileNetV2(weights="imagenet")
+    # Users can send HTTP requests with an image. The classifier will return
+    # the top prediction.
+    # Sample output: {"prediction":["n02099601","golden_retriever",0.17944198846817017]}
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py:21]:
```suggestion
    # Sample output: {"prediction":["n02099601","golden_retriever",0.17944198846817017]}
    
```

File: mobilenet/mobilenet.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py]: Could you reorder these two imports, so the longer one goes last:

```python
import tempfile
import numpy as np
```

File: mobilenet/mobilenet.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+from tensorflow.keras.preprocessing import image
+from tensorflow.keras.applications.mobilenet_v2 import (
+    preprocess_input,
+    decode_predictions,
+    MobileNetV2,
+)
+
+from ray import serve
+
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet.py:12]:
```suggestion


```

File: mobilenet/mobilenet_req.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import requests
+
+# TODO: Change this to your image path
+image_path = "/your/image/path"
```
Comment (shrekris-anyscale) [inline mobilenet/mobilenet_req.py:4]:
```suggestion
image_path = "/your/image/path"

```

**Review by shrekris-anyscale — APPROVED** [review]
Nice work! I left some minor comments on formatting. Once those are resolved, this change looks good to me.

**Comment by kevin85421** [comment]
@shrekris-anyscale By the way, which linter do you use? I can enable the linter on my IDE to avoid these issues in the future. Thanks!

**Comment by kevin85421** [comment]
Hi @shrekris-anyscale, I have already run linter [Black](https://marcobelo.medium.com/setting-up-python-black-on-visual-studio-code-5318eba4cd00#:~:text=Go%20to%20settings%20in%20your,%E2%80%9D%20and%20select%20%E2%80%9Cblack%E2%80%9D.) for all files. Thanks!

**Review by shrekris-anyscale — APPROVED** [review]
Awesome, thanks for addressing all the feedback @kevin85421! This change looks good to me.

**Pull request merged**